### ISSUE TITLE
Nmap doesn't skip using the hosts file when --dns-servers is used.

### ISSUE THREAD

**Issue by dariosusman** [nmap/nmap#2477]
I'm trying to scan a website for its SSL/TLS certificates, but I'm trying to make nmap skip using the /etc/hosts to resolve the FQDN using both and or --system-dns and --dns-servers <DNS SERVER>

**To Reproduce**
nmap  --system-dns -v -A <HOSTNAME>
nmap  --system-dns --dns-servers <DNS RESOLVER> -v -A <HOSTNAME>

I'd expect nmap NOT TO USE /etc/hosts if I specify it to use the system's resolver instead, or if I specify nmap a specific DNS resolver.

 - OS: Debian 11.3

 - Output of `nmap --version`: 
Nmap version 7.80 ( https://nmap.org )
Platform: x86_64-pc-linux-gnu
Compiled with: liblua-5.3.3 openssl-1.1.1j libssh2-1.9.0 libz-1.2.11 libpcre-8.39 libpcap-1.10.0 nmap-libdnet-1.12 ipv6
Compiled without:
Available nsock engines: epoll poll select

Thank you.

Best regards,
Dario Susman

**Comment by dmiller-nmap** [comment]
In addition to the points @nnposter raised (the system resolver, `getaddrinfo`, is used for forward lookups, and the `dns` options are for reverse only), there are a couple other points to consider:

1. Nmap *does* parse the `hosts` file and uses it to avoid doing reverse lookups for matching addresses.
2. The system resolver also uses the `hosts` file for forward (`getaddrinfo`) and reverse (`getnameinfo`) lookups, so `--system-dns` will always use the `hosts` file, regardless of what other changes we might make.

It may be useful in some cases to instruct Nmap to avoid using the system's `hosts` file, or to use a custom `hosts` file. We have the `--servicedb` file for specifying an alternative to `nmap-services` (with fallback to `/etc/services`), so a similar `--hostsdb` option could serve both these cases (either using `/dev/null` or a special value like `0` to mean "no lookups"). This would be especially useful if/when we enable Nmap to use its parallel DNS resolver for *forward* lookups (see #1451).

**Comment by nnposter** [comment]
If you only care about using some custom DNS server, without messing with your local DNS configuration, you could consider something like this:
```
export MYDNSSERVER=8.8.8.8
for hostname in `cat hostnames_to_scan`; do
    nmap -n -v -A --script-args http.host=$hostname,tls.servername=$hostname `dig @$MYDNSSERVER -4 +short $hostname | head -n 1`
done
```

**Comment by dariosusman** [comment]
> Please check Nmap documentation for the meaning of options `--system-dns` and `--dns-servers`. Namely:
> 
> > The system resolver is **always** used for forward lookups (getting an IP address from a hostname).
> 
> and
> 
> > --dns-servers server1[,server2[,...]]  (Servers to use for **reverse** DNS queries)
> 
> and
> 
> > This option _[--dns-servers]_ is **not honored** if you are using --system-dns.
> 
> Therefore these options will not accomplish what you want.
> 
> AFAICT, Nmap does not explicitly use `/etc/hosts` for forward look-ups. It is likely getting consulted by your system resolver.

Yeah, it normally does because it follows nsswitch (hosts dns), so it make sense. I just want nmap to just use DNS for forward-resolving.

Thank you!

Best regards,
Dario Susman
